### ISSUE TITLE
Tru-er SPA mode

### ISSUE THREAD

**Issue by ebeloded** [sveltejs/kit#1650]
SvelteKit allows building Single Page Apps with the following config:
```js
const adapter = require('@sveltejs/adapter-static');

module.exports = {
  kit: {
    adapter: adapter({
      fallback: '200.html'
    },
    prerender: {
      enabled: false
    },
    ssr: false
  }
};
```

**Is your feature request related to a problem? Please describe.**
Even with the `ssr` disabled in config file, it is impossible to reference `window` or `document` without wrapping it in an environment check. This is because each component needs to be evaluated to determine whether it has SSR export, as @Rich-Harris explained [here](https://github.com/sveltejs/kit/issues/754#issuecomment-810576165):
>The fundamental issue is that the component must be imported in order to determine whether it has an ssr export. That would continue to be true if we checked each page for ssr at build time and added the information to a manifest.

>There's a couple of options that spring to mind - we find some other way of establishing whether a given page should be SSR'd, or we determine it with static analysis (which has the usual caveats but should be pretty robust in this case).

**Describe the solution you'd like**

I'd like to have a way of configuring SvelteKit as a true SPA, in which I need not worry about referencing browser globals.

As a possible way forward, @benmccann [suggested](https://github.com/sveltejs/kit/issues/754#issuecomment-811176398) to expand the `ssr` option, allowing not just a boolean value:
> The other option would be to make the setting more than just a boolean. E.g. we could have true/false/per-page or something like that or always/never/true-by-default/false-by-default or whatever makes sense

I think this would be a fine solution.

**Describe alternatives you've considered**

The only alternative I currently see is to preface browser globals with environment check and load naughty dependencies dynamically.

**How important is this feature to you?**
Slightly annoying. Gets in the way. Forces to add meaningless code.

**Additional context**
The original issue regarding SPA in SvelteKit https://github.com/sveltejs/kit/issues/754

**Comment by umanghome** [comment]
Since `load` runs on the browser in SPA mode, it makes sense to be able to use `localStorage`.

**Comment by katendeglory** [comment]
It's really frustrating. 

I thought this could have been taken into account during the creation of sveltekit. 

Given the fact that we might want to use **browser globals** in a code that's meant to run on the **browser**.

What's the point of **SSR false** if we can't use local storage, document, the window object...?

**Comment by katendeglory** [comment]
What's the point of SSR false if we still can't use browser globals without hacks?

**Comment by einarpersson** [comment]
+1

I’d expect it to be possible to import `pixi.js` when I run SvelteKit in SPA-mode, but right now I have to do a dynamic import `await import(“pixi.js”)` because Pixi references the window object.

**Comment by Rich-Harris** [comment]
Reopening since this isn't actually completely fixed by #2008 — the page's module is still imported, and in a simple test app the server blows up when trying to render a placeholder for a page that imports a client-only dependency

**Comment by lemmon** [comment]
> Since `load` runs on the browser in SPA mode, it makes sense to be able to use `localStorage` from within `load`.

I think there's no need for checking valid variables/libraries. You should be able to reference any browser objects, without warnings, when `ssr` is set to `false`.

**Comment by brgrz** [comment]
Agree, this is an issue that still results in an compile error with `ssr: false` in svelte-kit config, for example

```
Error when evaluating SSR module /src/routes/...
ReferenceError: localStorage is not defined
```

The app still builds and runs locally but my Render builds are blocked by this.

**Comment by kilianso** [comment]
I'm having a hard time because of this using Firebase v9. Even with `ssr: false` and the `browser` check, my build fails using the static-adapter. I need to load all of their modules async/dynamically in order to build the app.

**Comment by didierfranc** [comment]
I'm doing that kind of hack waiting a better solution

```typescript
if (typeof window != 'undefined') {
	this.auth = getAuth(initializeApp(config));
	this.auth.onAuthStateChanged(function (currentUser) {
		if (currentUser) {
			authenticationStatus.set(AUTHENTICATION_STATUS.AUTHENTICATED);
			user.set(currentUser);
		} else {
			authenticationStatus.set(AUTHENTICATION_STATUS.UNAUTHENTICATED);
			user.set(null);
		}
	});
}
```

**Comment by kilianso** [comment]
@didierfranc I'm doing it by using dynamic imports and separated everything firebase-related into another JS file:

```
onMount(() => {
	(async () => {
		const {default: firebase} = await import('$lib/utils/firebase');
		firebase();
	})();
})
```
But even then, and even together with the `browser` check, if i do something like this: `import {getFirestore} from "firebase/firestore"`in this dynamically imported JS file, my build still fails with the same erors like the packages would have been imported regularly. I need to load everything dynamically in order to build the app.

**Comment by AndrewEQ** [comment]
I think it would be great to have a more "inline option" like this:

`
import client { chart } from 'svelte-apexcharts';
`

Default to existing experience (client & server) unless "client" or "server" is specified.

**Comment by dummdidumm** [comment]
That specific syntax is out of scope because it would enhance the JavaScript syntax, which is a no-go. There's a plugin though which does this by adding `?client` or `?server` to the end of the import string: https://github.com/bluwy/vite-plugin-iso-import

**Comment by AKuederle** [comment]
If you have imports that are only available on the client, it is very likely that you have entire parts of your javascript that is only expected to run on the client. Would it be feasible to have two separate script tags, one that would only be executed on the client and one that would be executed during SSR?

**Comment by AndrewEQ** [comment]
@dummdidumm yeah I'm currently doing it that way but yes, no editor support and you gotta install plugin and config it... not exactly an intuitive developer experience (DX) and I don't think anyone is expecting something so foundational to be found in the FAQ as well.

@AKuederle I wouldn't mind that either; at least you have a definitive way to separate the execution but it would be great if the DX were seamless... 

i.e. without you having to say that the library should be loaded or used in either environment and thereby alleviating the need to define another script tag, so if the library could "automagically" know that it could be loaded via server-side, client-side or both on its own which in a sense is currently happening, just that when something goes wrong, we need the option to disable one or the other. Having said that, perhaps some may want more fine-grain control which is why I suggested it on the import level vs script tag level. Now if the other shimmy with the `?client` would have worked with the editor, it would have been fine as well but just not the greatest DX.

**Comment by frederikhors** [comment]
I opened [this issue](https://github.com/sveltejs/kit/issues/2876) (closed by @bluwy) because I think we need to consider this too:

### Describe the bug

I couldn't understand why Kit builds my project two times even if [I disabled `ssr` in svelte.config.cjs](https://github.com/frederikhors/iss-svelte-kit-tailwind-comments/blob/master/svelte.config.js#L18):

```
vite v2.6.14 building for production...
...
vite v2.6.14 building SSR bundle for production...
```


### Reproduction

https://github.com/frederikhors/iss-svelte-kit-tailwind-comments

### Logs

```shell
> my-app@0.0.1 build
> svelte-kit build

vite v2.6.14 building for production...
✓ 28 modules transformed.
.svelte-kit/output/client/_app/assets/svelte-logo-87df40b8.svg             1.85 KiB
.svelte-kit/output/client/_app/manifest.json                               1.82 KiB
.svelte-kit/output/client/_app/pages/__layout.svelte-57ab7eca.js           4.04 KiB / gzip: 1.79 KiB
.svelte-kit/output/client/_app/error.svelte-0ebcf88e.js                    1.56 KiB / gzip: 0.75 KiB
.svelte-kit/output/client/_app/start-c23624a4.js                           18.75 KiB / gzip: 6.89 KiB
.svelte-kit/output/client/_app/pages/index.svelte-7be39bf6.js              3.62 KiB / gzip: 1.62 KiB
.svelte-kit/output/client/_app/pages/about.svelte-0cea0aea.js              2.29 KiB / gzip: 1.04 KiB
.svelte-kit/output/client/_app/pages/todos/index.svelte-d9cac2e2.js        5.48 KiB / gzip: 2.33 KiB
.svelte-kit/output/client/_app/assets/start-61d1577b.css                   0.16 KiB / gzip: 0.15 KiB
.svelte-kit/output/client/_app/chunks/vendor-836cc791.js                   14.11 KiB / gzip: 6.00 KiB
.svelte-kit/output/client/_app/assets/pages/__layout.svelte-fd9cb797.css   0.28 KiB / gzip: 0.19 KiB
vite v2.6.14 building SSR bundle for production...
✓ 31 modules transformed.
.svelte-kit/output/server/app.js                          0.10 KiB
.svelte-kit/output/server/chunks/app-32bc2016.js          51.76 KiB
.svelte-kit/output/server/chunks/index.json-784727b1.js   0.42 KiB
.svelte-kit/output/server/chunks/_api-e66b1c82.js         0.60 KiB
.svelte-kit/output/server/chunks/_uid_.json-039b6f30.js   0.41 KiB
.svelte-kit/output/server/chunks/__layout-d917862c.js     2.19 KiB
.svelte-kit/output/server/chunks/error-5c3307f4.js        0.75 KiB
.svelte-kit/output/server/chunks/index-b7efcc6a.js        6.08 KiB
.svelte-kit/output/server/chunks/about-950f4908.js        0.96 KiB
.svelte-kit/output/server/chunks/index-3b94195b.js        1.82 KiB

Run npm run preview to preview your production build locally.

> Using @sveltejs/adapter-static
  ✔ done
```


### System Info

```shell
System:
    OS: Windows 10 10.0.19043
  Binaries:
    Node: 16.13.0 - C:\Program Files\nodejs\node.EXE
    npm: 8.1.4 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: Spartan (44.19041.1266.0), Chromium (96.0.1054.29)
  npmPackages:
    @sveltejs/adapter-static: 1.0.0-next.21 => 1.0.0-next.21
    @sveltejs/kit: 1.0.0-next.196 => 1.0.0-next.196
    svelte: 3.44.2 => 3.44.2
```


### Severity

serious, but I can work around it

**Comment by IsaacHub** [comment]
It's very strange that @bluwy closed https://github.com/sveltejs/kit/issues/2876 without providing solutions. He asked to comment about it in this issue, but both are not related in anyway. 

Why build for SSR when it's set to false? Is this the intended behavior of sveltekit/vite bundling? 
Even though `fallback` is specified, `svelte-kit build` produces two bundles in sveltekit demo project using adapter-static.

How can we configure sveltekit to produce only client bundle and not ssr?

**Comment by frederikhors** [comment]
@bluwy can you please re-open https://github.com/sveltejs/kit/issues/2876? Thanks.

**Comment by bluwy** [comment]
There's no way to configure SvelteKit to not bundle for SSR without the effort to support true SPA mode first. That's why I requested the feature request to be added here as they are related. But sure let's bring the discussion to #2876.

**Comment by frederikhors** [comment]
https://github.com/sveltejs/kit/issues/2937

**Comment by surajbarkale** [comment]
After encountering numerous issues, here is the code that worked for me for firebase auth. I had to do the following:

1. Use Firebase SDK v9
2. Disable SSR using static adapter
3. Do not call any firebase functions in variable initialization
4. Import firebase types separately using `import type`

<details>
  <summary>firebase.ts</summary>

```typescript
import type { FirebaseApp } from "firebase/app";
import { initializeApp } from "firebase/app";

let app: FirebaseApp = null

export function firebaseApp() {
    if (app == null) {
        app = initializeApp({
            apiKey: "",
            authDomain: "",
            projectId: "",
            storageBucket: "",
            messagingSenderId: "",
            appId: "",
            measurementId: "",
        })
    }
    return app
}
```
</details>

<details>
  <summary>auth.ts</summary>

```typescript
import type { Readable, Subscriber, Unsubscriber } from 'svelte/store'
import { writable } from 'svelte/store'
import { firebaseApp } from './firebase'
import type { User, Auth } from '@firebase/auth'
import { initializeAuth, browserLocalPersistence, onAuthStateChanged } from "@firebase/auth";

let auth: Auth
let authLoaded = false
let isLoaded = writable(authLoaded)

export const isAuthLoaded = { subscribe: isLoaded.subscribe }

export function getAuth(): Auth {
    if (!auth) {
        auth = initializeAuth(firebaseApp(), {
            persistence: browserLocalPersistence,
        })
    }
    return auth
}

export async function getLoadedAuth(): Promise<Auth> {
    const auth = getAuth()
    if (authLoaded) {
        return auth
    }
    return await new Promise((resolve, reject) => {
        const unsubscribe = onAuthStateChanged(
            auth,
            _ => {
                authLoaded = true
                isLoaded.set(true)
                unsubscribe()
                resolve(auth)
            },
            reject
        )
    })
}

let userSubscribers = new Set<Subscriber<User>>()
let unsubscribeUser: Unsubscriber

export const currentUser: Readable<User> = {
    subscribe: (subscriber: Subscriber<User>): Unsubscriber => {
        userSubscribers.add(subscriber)
        const auth = getAuth()
        if (!unsubscribeUser) {
            unsubscribeUser = onAuthStateChanged(auth, user => {
                for (const subscriber of userSubscribers) {
                    subscriber(user)
                }
                if (!authLoaded) {
                    authLoaded = true
                    isLoaded.set(true)
                }
            })
        }
        subscriber(auth.currentUser)
        return () => {
            userSubscribers.delete(subscriber)
            if (userSubscribers.size == 0) {
                unsubscribeUser()
                unsubscribeUser = null
            }
        }
    }
}
```
</details>

**Comment by Rich-Harris** [comment]
hey all — flagging this discussion, as people here might have thoughts on it: https://github.com/sveltejs/kit/discussions/7716
